app/src/main/java/com/keylesspalace/tusky/components/notifications/PushNotificationHelper.kt

```diff
@@ -0,0 +1,216 @@
+suspend fun registerUnifiedPushEndpoint(context: Context, api: MastodonApi, accountManager: AccountManager, account: AccountEntity, endpoint: String) {
+    // Generate a prime256v1 key pair for WebPush
+    // Decryption is unimplemented for now, since Mastodon uses an old WebPush
+    // standard which does not send needed information for decryption in the payload
+    // This makes it not directly compatible with UnifiedPush
+    // As of now, we use it purely as a way to trigger a pull
+    val keyPair = CryptoUtil.generateECKeyPair(CryptoUtil.CURVE_PRIME256_V1)
+    val auth = CryptoUtil.secureRandomBytesEncoded(16)
+
+    withContext(Dispatchers.IO) {
+        api.subscribePushNotifications(
+            "Bearer ${account.accessToken}", account.domain,
+            endpoint, keyPair.pubkey, auth,
+            buildSubscriptionData(context, account)
+        ).onFailure {
+            Log.d(TAG, "Error setting push endpoint for account ${account.id}")
+            Log.d(TAG, Log.getStackTraceString(it))
+            Log.d(TAG, (it as HttpException).response().toString())
+
+            disableUnifiedPushNotificationsForAccount(context, account)
+        }.onSuccess {
+            Log.d(TAG, "UnifiedPush registration succeeded for account ${account.id}")
+
+            account.pushPubKey = keyPair.pubkey
+            account.pushPrivKey = keyPair.privKey
+            account.pushAuth = auth
+            account.pushServerKey = it.serverKey
+            account.unifiedPushUrl = endpoint
+            accountManager.saveAccount(account)
+        }
+    }
+}
+
+// Synchronize the enabled / disabled state of notifications with server-side subscription
+suspend fun updateUnifiedPushSubscription(context: Context, api: MastodonApi, accountManager: AccountManager, account: AccountEntity) {
+    withContext(Dispatchers.IO) {
+        api.updatePushNotificationSubscription(
+            "Bearer ${account.accessToken}", account.domain,
+            buildSubscriptionData(context, account)
+        ).onSuccess {
+            Log.d(TAG, "UnifiedPush subscription updated for account ${account.id}")
+
+            account.pushServerKey = it.serverKey
+            accountManager.saveAccount(account)
+        }
+    }
+}
+
+suspend fun unregisterUnifiedPushEndpoint(api: MastodonApi, accountManager: AccountManager, account: AccountEntity) {
+    withContext(Dispatchers.IO) {
+        api.unsubscribePushNotifications("Bearer ${account.accessToken}", account.domain)
+            .onFailure {
+                Log.d(TAG, "Error unregistering push endpoint for account " + account.id)
+                Log.d(TAG, Log.getStackTraceString(it))
+                Log.d(TAG, (it as HttpException).response().toString())
+            }
+            .onSuccess {
+                Log.d(TAG, "UnifiedPush unregistration succeeded for account " + account.id)
+                // Clear the URL in database
+                account.unifiedPushUrl = ""
```

Comment:
do we need to clear other parts?

Reply:
We don't really need to -- but just to be safe I changed it to clear everything